Add: jsx-no-unbounded-this-in-props

umurkontaci opened the pull request:
Disallows the usage of unbound callbacks to methods of `this` in ES6 style React classes.

I've tried to look for all the available usages and I think I have found all the ones that are in our codebase, running it in our codebase returned no errors. However, I know that it comes up from time to time when someone forgets to include the binding call. With this, we'll be able to capture them before they get in the wild.

/cc: @aduth 

Testing:
```bash
export WP_CALYPSO_DIR= # set this to correct value
export ESLINT_CALYPSO_DIR= # set this to correct value
cd "$ESLINT_CALYPSO_DIR"
git checkout add/jsx-no-unbounded-this-in-props
cd "$WP_CALYPSO_DIR"
find client -name '*.jsx' -or -name '*.js' | \
  xargs node_modules/.bin/eslint \
    --rule '{"jsx-no-unbound-members-in-props": 2}' \
    --rulesdir "$ESLINT_CALYPSO_DIR/lib/rules" \
    --no-eslintrc \
    --parser-options ecmaVersion:6 \
    --parser-options sourceType:module \
    --parser babel-eslint \
    --quiet
```

umurkontaci committed:
Add: jsx-no-unbounded-this-in-props

blowery commented on lib/rules/jsx-no-unbounded-this-in-props.js:
```diff
@@ -0,0 +1,93 @@
+	} );
+}
+const rule = {
+	ERROR_MESSAGE: 'JSX props should not unbounded `this`',
```
maybe 'JSX props should not use class members not bound to `this`. ?

blowery commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "test": "npm run tests && npm run lint"
   },
   "dependencies": {
+    "lodash": "4.17.4",
```
probably want this to be more general, so it can reuse the lodash already bundled by wp-calypso. maybe "^4.0"?

umurkontaci replied:
Ha, I've never thought of it that way, makes sense.

blowery commented on README.md:
```diff
@@ -53,6 +53,7 @@ Then configure the rules you want to use under the rules section.
 - [`jsx-gridicon-size`](docs/rules/jsx-gridicon-size.md): Enforce recommended Gridicon size attributes
 - [`import-docblock`](docs/rules/import-docblock.md): Enforce external, internal dependencies docblocks
 - [`post-message-no-wildcard-targets`](docs/rules/post-message-no-wildcard-targets.md): Disallow using the wildcard '*' in `postMessage`
+- [`jsx-no-unbounded-this-in-props`](docs/rules/jsx-no-unbounded-this-in-props.md): Disallow unbound `this` in JSX Props
```
maybe `jsx-no-unbound-members-in-props`? it's not really about `this` being unbound, it's about members being unbound.

jsnajdr commented:
> I've tried to look for all the available usages and I think I have found all the ones that are in our codebase, running it in our codebase returned no errors.

Hi @umurkontaci ! I wrote a similar script using the `jscodeshift` codemod library recently and it uncovered a few errors in the codebase:

https://github.com/Automattic/wp-calypso/pull/14040
https://github.com/Automattic/wp-calypso/pull/14037
https://github.com/Automattic/wp-calypso/pull/14036

(see my P2 post at p4TIVU-6zy-p2 for more info)

If your ESlint rule really didn't find them, can you check why?

umurkontaci commented:
> If your ESlint rule really didn't find them, can you check why?

Thanks for the ping, apparently there was a bug of how I test the rules. I fixed it and it has a couple false positive that needs to be fixed. Pulling it back to In Progress now.

umurkontaci committed:
Rename and fix false positives

umurkontaci commented:
Fixed some false positives and renamed things as @blowery suggested.

It now shows 20 errors in `wp-calypso`, and I hand checked all of them and confirmed they are actually bugs.

Output:
```
➜  wp-calypso git:(master) ✗ git show HEAD --oneline | head -n1
e5d958558 Comments: Detail view (#14004)
➜  wp-calypso git:(master) ✗ find client -name '*.jsx' -or -name '*.js' | xargs node_modules/.bin/eslint --rule '{"jsx-no-unbound-members-in-props": 2}' --rulesdir ../eslint-plugin-wpcalypso/lib/rules --no-eslintrc --parser-options ecmaVersion:6 --parser-options sourceType:module --parser babel-eslint --quiet

/Users/umurkontaci/Projects/wp-calypso/client/blocks/comments/comment-edit-form.jsx
  171:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  173:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  174:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  176:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  184:8   error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props

/Users/umurkontaci/Projects/wp-calypso/client/blocks/comments/form.jsx
  208:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  210:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  211:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  213:10  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  221:8   error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props

/Users/umurkontaci/Projects/wp-calypso/client/blocks/comments/post-comment.jsx
  105:53  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  175:31  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  181:70  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  215:6   error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props

/Users/umurkontaci/Projects/wp-calypso/client/my-sites/upgrades/cart/cart-plan-ad.jsx
  52:16  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props

/Users/umurkontaci/Projects/wp-calypso/client/reader/discover/post-attribution.jsx
  74:7  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  83:7  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  92:9  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props

/Users/umurkontaci/Projects/wp-calypso/client/reader/discover/site-attribution.jsx
  66:7  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props
  75:8  error  JSX props should not use class members not bound to `this`  jsx-no-unbound-members-in-props

✖ 20 problems (20 errors, 0 warnings)
```

klimeryk commented:
Awesome findings, @umurkontaci :1st_place_medal: Although, the ones in `blocks/comments/*` do seem to be false positives. Those classes have some [boilerplate code in the constructor](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/comments/form.jsx#L40-L44) to bind every `handle*` method:
```
		// bind event handlers to this instance
		Object.getOwnPropertyNames( PostCommentForm.prototype )
			.filter( ( prop ) => prop.indexOf( 'handle' ) === 0 )
			.filter( ( prop ) => typeof this[ prop ] === 'function' )
			.forEach( ( prop ) => this[ prop ] = this[ prop ].bind( this ) );
```

klimeryk commented:
And the ones in `client/reader/discover/site-attribution.jsx` seem to have just been fixed in https://github.com/Automattic/wp-calypso/pull/14141 😆
